### ISSUE TITLE
Calling grain method using GrainClient hangs after running for some time

### ISSUE THREAD

**Issue by chris-eaton** [dotnet/orleans#3830]
I am using ASP.NET and I have a controller method uses `GrainClient` within   If I run ASP.NET then instantly call this controller, the call to the grain works correctly.  

At some point in the future this stops working, I havent tracked down exactly how long it takes but certainly if I leave it running over night it fails in the morning.

If I step through the code I get a grain reference of  `{GrainReference:*grn/A97D1470/00000000}`. 

If I turn on all exceptions the call to the grain throws this exception

> System.Net.Sockets.SocketException: 'An existing connection was forcibly closed by the remote host'

The NativeErrorCode is `10054`

Oddly enough this exception doesnt seem to bubble up through ASP.NET and make the call fail.

**Comment by sergeybykov** [comment]
I'm not sure I understand. What do you mean by:
>If I turn on all exceptions 

Are you awaiting the grain calls? If you do, then every failed call should throw an exception. Is this not what you are seeing?

**Comment by chris-eaton** [comment]
Sorry, what I meant was no exceptions were being thrown so I turned on "Catch all exceptions" in the exceptions options of Visual Studio. But still no exceptions were thrown.

One thing I had noticed in another issue was that timeouts for Grain calls are quite large when ran under Debug? Is this true?

Here is my code for full transparency

```      
public async Task SaveRouteTemplate(RouteTemplate routeTemplate)
{
    if (this.bifrostClient.GrainClient == null)
    {
        throw new Exception("Bifrost client not started");
    }

    var geocodeGrain = this.bifrostClient.GrainClient.GetGrain<IGeocodeGrain>(0);

    foreach (var stop in routeTemplate.RouteStops)
    {
        try
        {
            stop.GeocodedAddress = await geocodeGrain.GetGeocode(stop.CenterPoint.Latitude, stop.CenterPoint.Longitude);
            stop.GeocodedAddressId = stop.GeocodedAddress.Id;
        }
        catch (Exception e)
        {
            Console.WriteLine(e);
        }
    }

    await this.tservice.AddRouteTemplate(routeTemplate);
}
```

**Comment by chris-eaton** [comment]
After probably 15-30 minutes the call finally timed out with this message

> Response did not arrive on time in 00:30:00 for message: Request *cli/b9620d92@928a4fa0->S192.168.5.34:40000:250937046*grn/A97D1470/00000000 #256: global::Grains.Interfaces.IGeocodeGrain:GetGeocode(). Target History is: <S192.168.5.34:40000:250937046:*grn/A97D1470/00000000:>.

I suppose I could change this timeout to be smaller, but the core issue for me here is that the client has disconnected.  It was my understanding that a client would auto reconnect if a connection was lost? Also during this time I am certain the Orleans network did not go down.  According to the Orleans dashboard each silo has been running for '3 weeks, 4 days'.

My client initialisation code is this:

```
public async Task<IClusterClient> InitialiseClient(ClientConfiguration config)
{
    config.AddSimpleMessageStreamProvider("SMSProvider");
    config.ClientName = System.Diagnostics.Process.GetCurrentProcess().ProcessName;

    config.TraceToConsole = this.TraceToConsole;

    var descriptionString = config.GatewayProvider == ClientConfiguration.GatewayProviderType.SqlServer
                    ? config.DataConnectionString
                    : "Local";

    Console.WriteLine($"Starting Client -> {descriptionString}");
    await this.orleansClientService.OnClientDebugMessage($"Starting Client -> {descriptionString}");

    IClusterClient clusterClient = null;

    while (clusterClient == null)
    {
        try
        {
            clusterClient = new ClientBuilder()
                .AddClusterConnectionLostHandler((o, e) =>
                {
                    ConnectionToClusterLost?.Invoke();
                })
                .UseConfiguration(config)
                .Build();

            await this.orleansClientService.OnClientDebugMessage($"Connecting");
            await clusterClient.Connect();
        }
        catch (SiloUnavailableException ex)
        {
            await this.orleansClientService.OnClientConnectionError(ex, $"Silo unavailable. {ex.Message}");

            this.LastException = ex;
        }
        catch (OrleansException ex)
        {
            await this.orleansClientService.OnClientConnectionError(ex, $"Orleans exception. {ex.Message}");
            this.LastException = ex;
        }
        catch (Exception ex)
        {
            await this.orleansClientService.OnClientConnectionError(ex, $"Unknown exception. {ex.Message}");
            this.LastException = ex;
        }
        finally
        {
            await this.orleansClientService.OnClientDebugMessage($"Waiting 1 second before trying again");
            await Task.Delay(TimeSpan.FromSeconds(1));

            if (clusterClient != null && !clusterClient.IsInitialized)
            {
                clusterClient.Dispose();
                clusterClient = null;
            }
        }
    }

    await this.orleansClientService.OnClientDebugMessage($"Client Started!");
    Console.WriteLine("Client Started!");

    return clusterClient;
}
```

**Comment by veikkoeeva** [comment]
@chris-eaton I don't have a reason for your problem and maybe you know this already, but maybe this is an idea worth implementing:

1. Write a wrapper that containes the client initialization logic and hands out connected cluster client instance when asked.
2. Add this wrapper as a singleton to the ASP.NET DI system.
3. Inject the wrapper to the controllers, ask a connected client from the wrapper.

If it's determined the connection is broken, ask a connected client from the wrapper. The wrapper in turn then connects again to the silo and returns a connected client. Depending of the situation, the controller can also just flag an error (and there can be a cancellation too if things take too long). This could potentially lower startup time as the Orleans connection logic can run on the background (inside the wrapper) when the system is starting otherwise too.

**Comment by chris-eaton** [comment]
@veikkoeeva Thanks for the response! It seems I do much of what you suggested already which reassures me that I cannot be too far away from a good solution.

So my flow is almost identical
- I have a ClientService that generates clients of a particular type
- The ClientService handles all state issues and calls callbacks on the Client
- This is a singleton injected into the ASP.NET DI system

i.e.
```builder.Register(context => new MyClient()).SingleInstance();
            builder.Register(context => new MyClientService<MyClient>(context.Resolve<MyClient>(), "My Client")).SingleInstance();
```

I was lead to believe that the ClusterClient should reconnect automatically.  You suggest that this should be a manual process? If so how can I achieve this? I already have a handler for the `ConnectionToCluster` lost event, I'm just unsure what to do in this case

```
clusterClient = new ClientBuilder()
    .AddClusterConnectionLostHandler((o, e) =>
    {
        ConnectionToClusterLost?.Invoke();
    })
    .UseConfiguration(config)
    .Build();
```

Thanks
Chris

**Comment by xiazen** [comment]
cluster client doesn't auto-reconnect if connection is lost. 
You can use `AddClusterConnectionLostHandler` to register a handler to handle the connection lost case. In the handler, gracefully disconnect your client by calling `IClusterClient.Close()` and reconnect manually.

**Comment by sergeybykov** [comment]
>cluster client doesn't auto-reconnect if connection is lost.

Client *does* reconnect upon a socket connection drop and reconfiguration of gateways. The two things that might get lost in certain failure conditions or a restart of the cluster are client stream subscriptions and client observers. Other than that, client should be able to continue communicating with the cluster even after a complete restart of it.

**Comment by chris-eaton** [comment]
Thanks for the responses.

So I am using subscriptions but not for this specific piece of functionality.  This is simply a grain call.

In this case what might be my best option to fix the issue?

**Comment by benjaminpetit** [comment]
> One thing I had noticed in another issue was that timeouts for Grain calls are quite large when ran under Debug? Is this true?

> After probably 15-30 minutes the call finally timed out with this message

Yes, when the debugger is attached, the value for grain timeout is overridden and set to 30 minutes.

> It was my understanding that a client would auto reconnect if a connection was lost? 

It will! And the client will be connected to several gateways if possible, so if one fail it should be fairly transparent.

The client might "blacklist" for a while a gateway that is failing too often (something like 3 failed connections in a row).

`AddClusterConnectionLostHandler` is here to notify that for a moment, your client was totally disconnected to the cluster (0 open connection to gateways), and you might need to renew Observer subscriptions, it has nothing to do with regular grain calls.

Do you have the client logs?

**Comment by chris-eaton** [comment]
Thanks Ben

Unfortunately I don't because its running under IIS the GrainClient doesnt have permission to write the logs to the folder. Is there a way I can specify which folder it should write logs to?

**Comment by berdon** [comment]
@benjaminpetit 
Is there any chance we could update this to only default to 30m if Debugger.IsAttached and you haven't overridden the default via configuration? This would explain why we've been seeing grain client timeouts in production but were unable to reproduce it locally.

**Comment by sergeybykov** [comment]
>Is there a way I can specify which folder it should write logs to?

In <=v1.5.3, you can explicitly set a full path to the log file via `ClientConfiguration.TraceFileName`.

>t looks like it'd probably be easier to add a "DisableDebuggerOverrides" configuration in MessagingConfiguration than try to predicate on the value not being set. :/

I like this idea.

**Comment by sergeybykov** [comment]
Resolved via #4193.
